### PR TITLE
`constrained-generators`: Add `lookup_` for maps

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/src/Constrained.hs [hunk 1]
```diff
@@ -101,12 +101,16 @@ module Constrained (
...
+  lookup_,
+  flip_,
   left_,
   right_,
   toGeneric_,
   fromGeneric_,
   sizeOf_,
   length_,
+  cNothing_,
+  cJust_,
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained.hs:113]: These got moved from the Conway Instances to here?
Oh! I see, these are from Maybe, and the Conway ones are for StrictMaybe.

File: libs/constrained-generators/src/Constrained/Examples/Basic.hs [hunk 2]
```diff
@@ -307,3 +307,11 @@ chooseBackwards' = constrained' $ \xys xy ->
...
       forAll ys $ \y -> 0 <. y
   ]
+
+whenTrueExists :: Specification BaseFn Int
+whenTrueExists = constrained $ \x ->
+  whenTrue (x ==. 0) $
+    exists (\_ -> pure False) $ \b ->
+      [ not_ b
+      , not_ (not_ b)
+      ]
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Examples/Basic.hs:317]: Isn't  [ not_b, not_ (not_b)]  always inconsistent?
Maybe some comments to explain how this is supposed to be used.?

File: libs/constrained-generators/src/Constrained/Examples/Map.hs [hunk 3]
```diff
@@ -74,3 +75,18 @@ elemSpec = constrained' $ \k v m ->
...
+  ]
+
+specificElemConstraints :: Specification BaseFn Int
+specificElemConstraints = constrained $ \k ->
+  [ assert $ k `elem_` lit [1 .. 10]
+  , forAll' (lit $ Map.fromList [(i, i) | i <- [1 .. 10]]) $ \k' v ->
+      whenTrue
+        (k ==. k')
+        (v ==. 10)
+  ]
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Examples/Map.hs]: This is meant for some kind of test?   Might explain what you expect this to do?

**Review by TimSheard — APPROVED** [review]
The propagateSpec instance is a bit daunting. But finally I wrapped my head around it.

File: libs/constrained-generators/src/Constrained/Examples/Basic.hs (hunk 2 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Examples/Basic.hs:317]: It is always inconsistent - so this spec means `not_ (x ==. 0)`. Before I fixed things here this example was crashing.

File: libs/constrained-generators/src/Constrained/Examples/Map.hs (hunk 3 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Examples/Map.hs]: I've removed it because it's not super interesting.

**Pull request merged**